### ISSUE TITLE
Need help in Test Double with queries

### ISSUE THREAD

**Issue by bobc82** [kenjis/ci-phpunit-test#127]
Hi, i'm learning about unit testing with CodeIgniter and i would to ask some questions about testing queries with Mocks class.

I tried to implement the following class News_model with the method get_news_all() that returns all news data from table 'news' and get_news returns only title and text fields from the same.

```php
class News_model extends CI_Model
{
    public function __construct(){
        $this->load->database();
    }

    public function get_news_all()
    {
        $query=$this->db->get('news');
        $result=$query->result_array();
        return $result;
    }

    public function get_news()
    {
        $this->db->select('title, text');
        $this->db->from('news');
        $query=$this->db->get();
        $result=$query->result_array();
        return $result;
    }
```

After i tried to build a News_model_with_mocks_test for testing method get_news_all() and in this case test runs fine:

```php
class News_model_with_mocks_test extends TestCase
{
    public function setUp()
    {
        $this->resetInstance();
        $loader=$this->getMockBuilder('CI_Loader')->setMethods(['database'])->getMock();
        $loader->method('database')->willReturn($loader);
        $this->CI->load=$loader;
        if(!class_exists('CI_DB', false))
        {
            eval('class CI_DB extends CI_DB_query_builder {}');
        }
        $this->obj=new News_model();
    }

    public function test_1()
    {
        $result_array = [
            [
                    "id" => "1",
                    "title" => "News",
                    "slug" => "news",
                    "text" => "News",
            ],
            [
                    "id" => "2",
                    "title" => "News2",
                    "slug" => "news2",
                    "text" => "News2",
            ],
        ];
        $db_result=$this->getMockBuilder('CI_DB_result')->disableOriginalConstructor()->getMock();
        $db_result->method('result_array')->willReturn($result_array);
        $db = $this->getMockBuilder('CI_DB')->disableOriginalConstructor()->getMock();
        $db->expects($this->once())->method('get')->with('news')->willReturn($db_result);
        $this->obj->db=$db;
        $result=$this->obj->get_news_all();
        $this->assertEquals($result_array,$result);
    }
```

But i don't know how to do tests for the method get_news(), i tried something as this:

```php
public function test_1()
{
    $result_array2 = [
            [
                    "title" => "News",
                    "text" => "News",
            ],
            [
                    "title" => "News2",
                    "text" => "News2",
            ],
    ];
    $db_result=$this->getMockBuilder('CI_DB_result')->disableOriginalConstructor()->getMock();
    $db_result->method('result_array')->willReturn($result_array2);
    $db = $this->getMockBuilder('CI_DB')->disableOriginalConstructor()->getMock();
    $db->expects($this->once())->method('query')->with('select title,text from news')->willReturn($db_result);
    $this->obj->db=$db;
    $result=$this->obj->get_news();
    $this->assertEquals($result_array2,$result);
}
```

phpunit thows the following exception:
PHP Fatal error:  Call to a member function result_array() on a non-   object in /opt/lampp/htdocs/codeigniter/application/models/Users_model.php on line 21

I don't know how to test double with select queries! Thank you in advance for your answers.

**Comment by yoX64** [comment]
You can test the model without that mock. Real database, real data. I think you are complicating too much.

**Comment by bobc82** [comment]
From 'Code Igniter Testing Guide': "_if our models have a lot of logic to do something like manipulating data from the database, it would be better to write unit testing without the database. It also might be better to think about separating database access from our model classes._"
I don't understand when is better to test with the database or without the database...

**Comment by yoX64** [comment]
It really depends on the situation. There are two big cases that you might find yourself when testing models.

**When the model only works with the data from the database** 
In this case(which you find yourself right now) it's better to make an _integration test_, meaning that your model will talk with a real database and will work with real operations from which you can see outcomes.

**When the model has a lot of business logic**
When I say business logic, I mean field validations, data manipulation, etc. In this case, because you have more things than only data transations with the database, you must also test that the logic that is there is written correctly. When you have this situation it's best to have unit tests on the model(meaning that you will mock the database transactions) and integrations tests on the model(as explained in the first case).

Maybe it's a bit confusing that you expect from this plugin (no harm intended @kenjis)  to do unit test with the models, but actually that is integration test, because you test(at least) 2 components (the model and the class that works with the database). Unit test is when you only test one class and all the dependencies are mocked.

TL;DR
Because you only retrieve data from the database, use a real database. You can find how to do it in the [docs](https://github.com/kenjis/ci-phpunit-test/blob/master/docs/HowToWriteTests.md) of this plugin.

**Comment by bobc82** [comment]
I know how do it and it's simpler and more intuitive than mocks. But i didn't find so much documentation about mocks outside the examples that i read in Codeigniter Testing Guide. I would be glad if you could advise me about documentation mode detailed about mocks. Thank you!

**Comment by yoX64** [comment]
[This tutorial](https://github.com/kenjis/ci-phpunit-test/blob/master/docs/HowToWriteTests.md#using-phpunit-mock-objects) will show you exactly how you can do it.

**Comment by bobc82** [comment]
Thank you again!

**Comment by kenjis** [comment]
@bobc82 I recommend you read *6.2 PHPUnit Mock Objects* (P.75-) in "CodeIgniter Testing Guide".
